lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2974,6 +2974,20 @@ ConformanceChecker::inferTypeWitnessesViaValueWitnesses(ValueDecl *req) {
                                          result.second->getCanonicalType()})
                              .second)
                          return true;
+                       
+                       // Filter out circular possibilities, e.g. that
+                       // AssocType == S.AssocType or
+                       // AssocType == Foo<S.AssocType>.
+                       if (result.second.findIf([&](Type t) -> bool {
+                         if (auto dmt = t->getAs<DependentMemberType>()) {
+                           if (dmt->getAssocType()->getName() == result.first->getName()
+                             && dmt->getBase()->isEqual(Conformance->getType()))
```

Comment:
I guess using `Conformance->getType()` here works because this is the result of a weird partial substitution; a `DependentMemberType` with a nominal type as its base isn't something that exists in any well-formed type.